pkg/controller/nodelink/nodelink_controller.go

```diff
@@ -0,0 +1,394 @@
+package nodelink
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	mapiv1beta1 "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	machineAnnotationKey = "machine.openshift.io/machine"
+)
+
+// Add creates a new Nodelink Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	reconciler := newReconciler(mgr)
+	return add(mgr, reconciler, reconciler.nodeRequestFromMachine)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) *ReconcileNodeLink {
+	return &ReconcileNodeLink{
+		client:              mgr.GetClient(),
+		scheme:              mgr.GetScheme(),
+		providerIDToMachine: make(map[string]*mapiv1beta1.Machine),
+		providerIDToNode:    make(map[string]*corev1.Node),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler, mapFn handler.ToRequestsFunc) error {
+	// Create a new controller
+	c, err := controller.New("nodelink-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	//Watch for changes to Node
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Machines and enqueue if it exists the backed node
+	err = c.Watch(&source.Kind{Type: &mapiv1beta1.Machine{}}, &handler.EnqueueRequestsFromMapFunc{ToRequests: mapFn})
```

Comment:
I don't think this is actually needed. Any node object is periodically updated (by kubelet updating node status). So, even if a machine is changed (e.g. labels added, taints added), referenced node will get eventually updated. By watching machines you are increasing the rate in which node requests are processed.

I was more interesting if there is a way to put two different kinds into the same queue and process them in either the same `Reconcile` method or in two distinct.

Reply:
Given the node heartbeat frequency this won't make a substantial difference, there might be cases where heartbeat fails and this pursues correctness where reconciliation wants to be triggered by a machine change, e.g taints/labels.
This is pretty much doing what you state in the last paragraph https://github.com/kubernetes-sigs/controller-runtime/blob/master/FAQ.md#q-how-do-i-know-which-type-of-object-a-controller-references